File: flutter_client/lib/features/trade/logic/ticket_qty.dart
/// 极速单仓位：A 股 100 股一手，港美 1 股。不按板手（港股每票不同）估算。
int lotSizeForMarket(String market) => market.toUpperCase() == 'CN' ? 100 : 1;

String cashCurrencyForMarket(String market) => switch (market.toUpperCase()) {
  'CN' => 'CNY',
  'HK' => 'HKD',
  _ => 'USD',
};

/// 按购买力 / 可卖数量把 25/50/75/100% 落成股数。
///
/// 买入：`floor(cash / price / lot) * lot` 再按百分比向下取整到手数。
/// 卖出：可卖 × 百分比，A 股再向下取整到 100。
/// 缺价格、现金或可卖时返回 0，由 UI 提示，不编造手数。
int ticketQtyForPercent({
  required int percent,
  required String side,
  required String market,
  double? price,
  double? cash,
  double? sellable,
}) {
  if (percent <= 0) {
    return 0;
  }
  final lot = lotSizeForMarket(market);
  final pct = percent.clamp(1, 100);
  if (side == 'sell') {
    final avail = sellable ?? 0;
    if (avail <= 0) {
      return 0;
    }
    final raw = (avail * pct / 100).floor();
    if (lot <= 1) {
      return raw;
    }
    return (raw ~/ lot) * lot;
  }
  if (price == null || price <= 0 || cash == null || cash <= 0) {
    return 0;
  }
  final maxShares = (cash / price / lot).floor() * lot;
  if (maxShares <= 0) {
    return 0;
  }
  return ((maxShares / lot) * pct / 100).floor() * lot;
}

File: ruoyi-fastapi-frontend/src/mobile/utils/ticketQty.js
/** Client-only lot size: A-share 100, HK/US 1. Never invent lots when price or buying power is missing. */

export function lotSizeForMarket(market) {
  return String(market || '').toUpperCase() === 'CN' ? 100 : 1
}

export function cashCurrencyForMarket(market) {
  const m = String(market || '').toUpperCase()
  if (m === 'CN') return 'CNY'
  if (m === 'HK') return 'HKD'
  return 'USD'
}

/**
 * Map 25/50/75/100% to a share count.
 * Buy: floor(cash / price / lot) * lot, then take percent of lots.
 * Sell: sellable * percent, A-share floored to 100.
 * Missing price / cash / sellable → 0 (UI shows a tip).
 */
export function ticketQtyForPercent({ percent, side, market, price, cash, sellable }) {
  if (!percent || percent <= 0) return 0
  const lot = lotSizeForMarket(market)
  const pct = Math.min(100, Math.max(1, Number(percent) || 0))
  if (side === 'sell') {
    const avail = Number(sellable)
    if (!Number.isFinite(avail) || avail <= 0) return 0
    const raw = Math.floor((avail * pct) / 100)
    if (lot <= 1) return raw
    return Math.floor(raw / lot) * lot
  }
  const px = Number(price)
  const buyingPower = Number(cash)
  if (!Number.isFinite(px) || px <= 0 || !Number.isFinite(buyingPower) || buyingPower <= 0) {
    return 0
  }
  const maxShares = Math.floor(buyingPower / px / lot) * lot
  if (maxShares <= 0) return 0
  return Math.floor(((maxShares / lot) * pct) / 100) * lot
}

export function inferMarket(symbol, fallback = 'US') {
  const s = String(symbol || '').toUpperCase()
  if (s.endsWith('.HK') || s.includes('.HK')) return 'HK'
  if (s.endsWith('.US')) return 'US'
  if (s.endsWith('.SH') || s.endsWith('.SZ') || s.endsWith('.SS')) return 'CN'
  const code = s.split('.')[0]
  if (/^\d{6}/.test(code)) return 'CN'
  if (/^\d{1,5}$/.test(code)) return 'HK'
  return fallback || 'US'
}

export function quoteSymbol(symbol) {
  return String(symbol || '').split('.')[0]
}
